frb_codegen/src/library/integration/integrator.rs

```diff
@@ -33,33 +43,83 @@ pub fn integrate(config: IntegrateConfig) -> Result<()> {
         .clone()
         .unwrap_or(format!("rust_lib_{}", dart_package_name));
 
-    extract_dir_and_modify(
-        &INTEGRATION_TEMPLATE_DIR,
+    info!("Overlay template onto project");
+    let mut replacements = HashMap::new();
+    replacements.insert("REPLACE_ME_DART_PACKAGE_NAME", dart_package_name.to_string());
+    replacements.insert("REPLACE_ME_RUST_CRATE_NAME", rust_crate_name.to_string());
+    replacements.insert("REPLACE_ME_RUST_CRATE_DIR", config.rust_crate_dir);
+    replacements.insert("REPLACE_ME_FRB_VERSION", env!("CARGO_PKG_VERSION").to_string());
+    let rust_frb_dependency = if config.enable_local_dependency {
+        r#"{ path = "../../../frb_rust" }"#.to_owned()
+    } else {
+        format!(r#""={}""#, env!("CARGO_PKG_VERSION"))
+    };
+    replacements.insert("REPLACE_ME_RUST_FRB_DEPENDENCY", rust_frb_dependency);
+    overlay_dir(
+        &SHARED_INTEGRATION_TEMPLATE_DIR,
+        &replacements,
         &dart_root,
-        &|path, src_raw, existing_content| {
+        &|target_path, reference_content, existing_content| {
             modify_file(
-                path,
-                src_raw,
+                target_path.into(),
+                reference_content,
                 existing_content,
-                &dart_package_name,
-                &rust_crate_name,
-                &config.rust_crate_dir,
+                &replacements,
                 config.enable_local_dependency,
+                None,
             )
         },
         &|path| filter_file(path, config.enable_integration_test),
     )?;
+    match &config.template {
+        Template::App => overlay_dir(
+            &APP_INTEGRATION_TEMPLATE_DIR,
+            &replacements,
+            &dart_root,
+            &|target_path, reference_content, existing_content| {
+                modify_file(
+                    target_path.into(),
+                    reference_content,
+                    existing_content,
+                    &replacements,
+                    config.enable_local_dependency,
+                    Some(&["main.dart"]),
+                )
+            },
+            &|path| filter_file(path, config.enable_integration_test),
+        )?,
+        Template::Plugin => overlay_dir(
+            &PLUGIN_INTEGRATION_TEMPLATE_DIR,
+            &replacements,
+            &dart_root,
+            &|target_path, reference_content, existing_content| {
+                modify_file(
+                    target_path.into(),
+                    reference_content,
+                    existing_content,
+                    &replacements,
+                    config.enable_local_dependency,
+                    Some(&[&format!("{}.dart", &dart_package_name)]),
+                )
+            },
+            &|path| filter_file(path, config.enable_integration_test),
+        )?,
+    }
```

Comment:
nit: The two blocks looks quite duplicated. maybe we can extract the common parts. e.g. easiest way is to

```
let (dir, somename) = match ... {
    App => (APP_INTEGRATION_TEMPLATE_DIR, "main.dart"),
   Plugin => ... 
};
...